compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/source/BeamUnboundedSourceVertex.java

```diff
@@ -0,0 +1,176 @@
+  }
+
+  @Override
+  public IRVertex getClone() {
+    return new BeamUnboundedSourceVertex<>(this);
+  }
+
+  @Override
+  public List<Readable<WindowedValue<O>>> getReadables(final int desiredNumOfSplits) throws Exception {
+    final List<Readable<WindowedValue<O>>> readables = new ArrayList<>();
+    source.split(desiredNumOfSplits, null)
+      .forEach(unboundedSource -> readables.add(new UnboundedSourceReadable<>(unboundedSource)));
+    return readables;
+  }
+
+  @Override
+  public void clearInternalStates() {
+    source = null;
+  }
+
+  @Override
+  public ObjectNode getPropertiesAsJsonNode() {
+    final ObjectNode node = getIRVertexPropertiesAsJsonNode();
+    node.put("source", sourceDescription);
+    return node;
+  }
+
+  /**
+   * UnboundedSourceReadable class.
+   * @param <O> output type.
+   * @param <M> checkpoint mark type.
+   */
+  private static final class UnboundedSourceReadable<O, M extends UnboundedSource.CheckpointMark>
+      implements Readable<WindowedValue<O>> {
+    private final UnboundedSource<O, M> unboundedSource;
+
+    UnboundedSourceReadable(final UnboundedSource<O, M> unboundedSource) {
+      this.unboundedSource = unboundedSource;
+    }
+
+    @Override
+    public Iterable<WindowedValue<O>> read() throws IOException {
+      return new UnboundedSourceIterable<>(unboundedSource);
+    }
+
+    @Override
+    public List<String> getLocations() throws Exception {
+      return null;
+    }
+  }
+
+  /**
+   * The iterable class for unbounded sources.
+   * @param <O> output type.
+   * @param <M> checkpoint mark type.
+   */
+  private static final class UnboundedSourceIterable<O, M extends UnboundedSource.CheckpointMark>
+      implements Iterable<WindowedValue<O>> {
+
+    private UnboundedSourceIterator<O, M> iterator;
+
+    UnboundedSourceIterable(final UnboundedSource<O, M> unboundedSource) throws IOException {
+      this.iterator = new UnboundedSourceIterator<>(unboundedSource);
+    }
+
+    @Override
+    public Iterator<WindowedValue<O>> iterator() {
+      return iterator;
+    }
+  }
+
+  /**
+   * The iterator for unbounded sources.
+   * @param <O> output type.
+   * @param <M> checkpoint mark type.
+   */
+  private static final class UnboundedSourceIterator<O, M extends UnboundedSource.CheckpointMark>
+      implements Iterator<WindowedValue<O>> {
+
+    private final UnboundedSource.UnboundedReader<O> unboundedReader;
+    private boolean available;
+
+    UnboundedSourceIterator(final UnboundedSource<O, M> unboundedSource) throws IOException {
+      this.unboundedReader = unboundedSource.createReader(null, null);
+      available = unboundedReader.start();
+    }
+
+    @Override
+    public boolean hasNext() {
+      // Unbounded source always has next element until it finishes.
+      return true;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public WindowedValue<O> next() {
+      try {
+        while (true) {
+          if (!available) {
```

Comment:
Where does this variable become true? If it is false, it loops infinitely.

Reply:
+1

Reply:
addressed